ugvc/mrd/bqsr_train_utils.py

```diff
@@ -0,0 +1,807 @@
+    df, max_score, label, score="qual", cumulative=False, apply_log_trans=True
+):
+    precisions = []
+    recalls = []
+    fprs = []
+    for i in range(max_score):
+        if cumulative:
+            d_q = df[df[score].apply(np.floor) >= i]
+        else:
+            d_q = df[df[score].apply(np.floor) == i]
+        precisions.append(get_precision(d_q, label))
+        recalls.append(get_recall(d_q, df, label))
+        if precisions[-1] == np.nan:
+            fprs.append(np.none)
+        elif apply_log_trans:
+            qual = -10 * np.log10(1 - precisions[-1])
+            qual = min(qual, max_score)
+            fprs.append(qual)
+        else:
+            fprs.append(precisions[-1])
+    return fprs, recalls
+
+
+def plot_precision_recall(lists, labels, max_score, log_scale=False):
+    for lst, label in zip(lists, labels):
+        plt.plot(lst[0:max_score], ".-", label=label)
+        plt.legend()
+        if log_scale:
+            plt.yscale("log")
+        plt.xlabel("QUAL")
+
+
+def bqsr_train(
+    hom_snv_featuremap,
+    single_substitution_featuremap,
+    out_path: str,
+    out_basename: str = "",
+    reference_fasta: str = None,
+    flow_order: str = DEFAULT_FLOW_ORDER,
+    training_chrom: str = "chr20",
+):
+    # init
+    os.makedirs(out_path, exist_ok=True)
+    if len(out_basename) > 0 and not out_basename.endswith("."):
+        out_basename += "."
+
+    params = {}    
+    params["workdir"] = out_path
+    params["hom_snv_featuremap"] = hom_snv_featuremap
+    params["single_substitution_file"] = single_substitution_featuremap
+    params["data_name"] = (
+        params["single_substitution_file"].split("/")[-1].split(".featuremap")[0]
+    )
+    params["ref_fasta"] = reference_fasta
+    params["training_chrom"] = training_chrom
+    params["flow_order"] = flow_order
+
+    params = model_params(params)
+
+    # load data
+    data = extract_data_from_featuremap_local(params)
+    print(data["label"].value_counts())
+
+    # xgb model: init and train
+    motif_only_features = [
+        "ref_alt_motif",
+        "left_motif",
+        "right_motif",
+    ]  
+
+    categorical_features = ["cycle_skip_status"] + motif_only_features
+    non_categorical_features = [
+        "X_SCORE",
+        "X_EDIST",
+        "X_LENGTH",
+        "X_MAPQ",
+        "X_INDEX",
+        "max_softclip_len",
+        "X_FC1",
+    ]
+
+    features = categorical_features + non_categorical_features
+    if "is_mixed" in data.columns:
+        features.append("is_mixed")
+
+    label_columns = ["label"]
+
+    for col in categorical_features:
+        data[col] = data[col].astype("category")
+
+    data = balance_motifs(data, params)
+    data["label"][data["label"] == "TP"] = 1
+    data["label"][data["label"] == "FP"] = 0
+    data = data.astype({"label": "int16"})
+
+    X_train, X_test, y_train, y_test = train_test_split(
+        data[features], data[label_columns], test_size=0.33, random_state=42
```

Comment:
I suggest making test_size a numerical value and not a ratio. IMO you don't generally need a third of the data as a test set, you just need a large enough test set to measure performance.

Reply:
decide training set size from all genome, ad decide the test set size